cobra/core/model.py

```diff
@@ -489,71 +483,58 @@ def to_array_based_model(self, deepcopy_model=False, **kwargs):
-
-        time_limit : Maximum solver time (in seconds)
-
-        .. NOTE :: Only the most commonly used parameters are presented here.
-                   Additional parameters for cobra.solvers may be available and
-                   specified with the appropriate keyword argument.
+        tolerance_feasibility : float
+            Solver tolerance for feasibility.
+        tolerance_markowitz : float
+            Solver threshold during pivot.
+        time_limit : float
+            Maximum solver time (in seconds).
+
+        Notes
+        -----
+        Only the most commonly used parameters are presented here.  Additional
+        parameters for cobra.solvers may be available and specified with the
+        appropriate keyword argument.
 
         """
-        current = interface_to_str(self.solver.interface.__name__)
-        so = kwargs.get('solver', 'optlang-' + current)
-        # after deprecation this can be checked with:
-        # if so in solvers:
-        if so in ('optlang-' + k for k in solvers):
-            if interface_to_str(so) != current:
-                self.solver = interface_to_str(so)
-            self._timestamp_last_optimization = time.time()
-            original_direction = self.solver.objective.direction
-            if objective_sense is not None:
-                self.solver.objective.direction = \
-                    {'minimize': 'min', 'maximize': 'max'}[objective_sense]
-            # Please note that the solution must always be extracted right
-            # after solver.optimize() since some solvers such as cplex
-            # invalidate their solution if the model is changed afterwards
-            self.solver.optimize()
-            # Not nice, but necessary until next optlang release
-            solution = solution_type(self)
-            # solution = (solution_type(self) if
-            #             self.solver.status == 'optimal' else self.solution)
-            if objective_sense is not None:
-                self.solver.objective.direction = original_direction
-        else:
+        (legacy, solver) = choose_solver(self, solver=kwargs.get("solver"))
+
+        if legacy:
             solution = optimize(self, objective_sense=objective_sense,
                                 **kwargs)
-        self.solution = solution
-
-        if solution.status is not 'optimal':
-            raise SolveError('no optimal solution')
-            # TODO: make failing optimization raise suitable exception
-            # raise exceptions._OPTLANG_TO_EXCEPTIONS_DICT.get(solution.status,
-            #                                                  SolveError)(
-            #     'Solving model %s did not return an optimal solution. The '
-            #     'returned solution status is "%s"' % (
-            #         self, solution.status))
+        else:
+            original_solver = self.solver
+            original_direction = self.solver.objective.direction
+            self.solver = solver
```

Comment:
What is managed differently? I agree that some of my code does not use some of the helper functions that were introduced later like `fix_objective_as_constraint`, etc but I'm still planning to fix that...
I know that FVA does not use the context manager as much since that resetting the contexts is slower than running the entire FVA...